### ISSUE TITLE
[UX] Design right-click behavior for Focus Mode

### ISSUE THREAD

**Issue by jonbell-lot23** [replayio/design#46]
**The problem we're trying to solve**

Focus Mode should be contextual. You should be able to be in a situation, right click on something, and automatically narrow the focus to that point. This is going to be a great feature.

**Things we know**

* It'll be a standard right-click iteraction
* It'll be used in multiple places
* It'll be rad

**The best way forward**

1. On the design side, I'm not sure exactly where we want to use this. I'll need help from the team defining entry points, so let's discuss that in the comments below.

2. Once we have the entry points figured out, I'll need help figuring out what options we want to see in the menu. "Focus here"? "Start point here" and "End point here?" We'll want to discuss as a group. I will bias towards fewer options, rather than a menu with too many options at first, but we'll see.

**Comment by jonbell-lot23** [comment]
Hey @bvaughn @jaril @jasonLaster @markerikson see notes above. Can we discuss these two things?

* Where should these entry points be?
* What would we expect to see in the menu?

I have my own idea -- print statements, timeline, console for entry points, and I want the menu to start simple. But would benefit from your thoughts.

**Comment by bvaughn** [comment]
Vague question: our client side modeling around focus region is all time based. I _think_ we're going to need to switch over to be _point based_ for some of the functionality Jon's describing above. We have a function (`mostRecentPaintOrMouseEvent`) that looks up the nearest paint event (with a point) for a given time – but can we easily/reliably make the opposite conversion? Will this change require dramatically re-working focus mode?

**Comment by jaril** [comment]
There's probably a budget version where we wouldn't need points. But this definitely feels like something where precision down to the point's important.

> We have a function (`mostRecentPaintOrMouseEvent`) that looks up the nearest paint event (with a point) for a given time – but can we easily/reliably make the opposite conversion? Will this change require dramatically re-working focus mode?

Not off the top of my head. @hbenl @jcmorrow?

**Comment by jaril** [comment]
Also I _think_ there might be a way to avoid having to do any point<>time conversions, but I'm not sure how janky it really is. We essentially have a list of points, and we're used the focus region start/end to filter it. There might be a version where that start/end value can be either a point OR time, depending on how the focus was set. The points have both a `point` and `time` so we can be flexible.

The problem that idea runs into though is if your focus start is a point, the (global) timeline has no concept of points. So that's where we would need some conversion to a time.

Yeah not sure how I like that but keeping track of that thought here.

**Comment by jonbell-lot23** [comment]
Adding comments from @jcmorrow 

>I think places that I often think about this are the breakpoint timelines and the console/network monitor (anything that has the pattern of a vertical panel with a horizontal red timeline that scrolls down it basically). So, setting a particular console message/network request as your start or end time (this gets weird for network requests because they are not a single instant in time but whatever, that's a detail). For breakpoint analysis, I've been trying to figure out what the right set of options is. Maybe it's  as simple as like "start at this hit", "end at this hit"? Also, side note, I could see Mark's idea of scrolling to zoom in/out working well on the breakpoint timeline.

**Comment by markerikson** [comment]
FWIW, I could maybe see a parallel to how the Beyond Compare diffing tool lets you select files to compare from within Windows Explorer.

It adds a context menu item for all files.  When you right-click any file, it first says "Select left file for compare":

![image](https://user-images.githubusercontent.com/1128784/168498193-6fd44acf-0bda-4bb8-8538-8675afa7f7ae.png)

When you choose that, BC saves the file path into the registry or something.

When you then right-click another file, BC changes the menu item and now says "Compare to `otherFile.ts`":

![image](https://user-images.githubusercontent.com/1128784/168498229-5c55b05a-de51-479d-81a8-22b6d79532f9.png)

I kinda like the idea of being able to select arbitrary items as the start and end this way.  They don't even have to be the _same_ kind of thing!  Like, you could do "Start focus at this breakpoint", and "End focus at this network request".

**Comment by jonbell-lot23** [comment]
Great point, @markerikson ! I didn't see these comments until I recorded the below video but the concepts I'm showing here absolutely work with the model you're describing.

cc @jasonLaster @jcmorrow @bvaughn @jaril check out this video, curious to hear your thoughts:
https://www.loom.com/share/227734dd67374c378114db4358e3c420

**Comment by jonbell-lot23** [comment]
[Here's the Figma I'm working from](https://www.figma.com/file/mAotNZRv6M5X0zHIFKPzFY/?node-id=12193%3A85238)

**Comment by jasonLaster** [comment]
I'm not sure. There is something nice about smart / cute... There is also something nice about explicit / learnable.

I think it might come down to copy. i'm in favor of something like either

Option 1
* Start focusing here
* Stop focusing here

Option 2
* Set focus start
* Set focus stop

Of the two I think i like option 2 because it is shorter and more direct. Also - i think iconagraphy could real help clarify as well. Think a modified focus icon with a blue triangle on the left or right.

**Comment by jonbell-lot23** [comment]
"Cute" versus "explicit" is the wrong framing, because we all agree explicit is better.

For the first click, it's a question of two approaches:

"Focus here"
vs
"Set focus start"

The first option does what the rest of the UX does -- it picks where you are and comes up with a reasonable range around it for you to fine-tune if you choose. It uses our existing model.

The second option causes you to always pick two points, the first one (start) and the second one (end). It also forces a different interaction model from the one we're already using in the timeline.

So the burden of proof is on why adding a new interaction model and extra clicks is worth it.

**Comment by jonbell-lot23** [comment]
Update:

1. I forgot that in the console you don't want a range. You want to start HERE.
2. Also we totally need the unicorn palette, duh

This is a quick mockup but it gets the general idea across:

![image](https://user-images.githubusercontent.com/9154902/168518174-6b3fb439-4d72-46ca-bdd8-4f8b024c938d.png)

Stay tuned, will improve these visuals

**Comment by jcmorrow** [comment]
> Also I think there might be a way to avoid having to do any point<>time conversions, but I'm not sure how janky it really is. We essentially have a list of points, and we're used the focus region start/end to filter it. There might be a version where that start/end value can be either a point OR time, depending on how the focus was set. The points have both a point and time so we can be flexible.
>
> The problem that idea runs into though is if your focus start is a point, the (global) timeline has no concept of points. So that's where we would need some conversion to a time.
>
> Yeah not sure how I like that but keeping track of that thought here.

We do deal with times without points, but that's partially why we've added `getPointNearTime`. For sure it would be possible to also write a `getTimeNearPoint`, but I don't know where we would need that, most points are TimeStampedPoints which will have a time attached.

**Comment by jcmorrow** [comment]
I dunno, on the whole "focus here" vs. "start and end" thing... I don't really think they are very different.

> The first option does what the rest of the UX does -- it picks where you are and comes up with a reasonable range around it for you to fine-tune if you choose. It uses our existing model.

I think the word "reasonable" is filling in a lot of gaps in this sentence. Reasonable when I'm looking at a network request which took several seconds is very different from reasonable when I'm looking at a tight loop that was hit 1,000 times in a second and I'm drowing in console logs. The UI for saying "focus here" is simple, but getting the "reasonable" range right might be really hard.

Re: extend vs. start and stop. This only makes sense when you are making the window *bigger*. Sure, if the user has a window already selected and they pick a point before the beginning, we are pretty sure they want to "start" there because ending there makes no sense. Likewise if they are choosing something after the end of the currently selected window. But it would be infuriating if when a user chose a point *inside* of the current window the UI just said "extend to here" or "shrink to here". If I were staring at that button, I would think "I have no idea what this button is going to do".

**Comment by jonbell-lot23** [comment]
@jcmorrow That totally makes sense, yeah. The "set start" and "set end" pattern above is looking better and better.

**Comment by bvaughn** [comment]
Wow, a lot of things to reply to. GitHub issues don't work well for this. Maybe we should use Discussions for this sort of thing so they are threaded! 😃

---

# Point vs Time

@jaril: I don't understand what you mean by this comment: https://github.com/RecordReplay/design/issues/46#issuecomment-1126150123

---

# Context Menu Loom

Regarding [this Loom](https://www.loom.com/share/227734dd67374c378114db4358e3c420) from @jonbell-lot23: I love the simplicity of "focus around this point" and the start/end refinements, but my gut is that they aren't always going to work and we shouldn't try to be smart about limiting the options like that. (For example in the video around 1:03, I don't think "start focusing here" is the obvious right choice; seems just as likely that the user might want to move the end point to there.) I do like having "focus around this point" as an option though. I think though that we should probably also give explicit start/stop options. Those seem like likely use cases to me (e.g. "I've found the start of the thing I care about. I'll move forward and find the end next.")

---

# Network Filtering

Big agree 👍  with the point @jcmorrow made about what's "reasonable" and the concern he raised about Network requests straddling a focus boundary– (e.g. request was made outside of the window but response came back during the window, or the other way around).

My immediate instinct was that if it overlapped at all, we should include it– but then I think that could lead to awkward edge cases where the UI could show an unfocused request in between two focused requests (maybe because an request sent earlier took longer to get a response).

Basically I _think_ our filtering logic should match our display/sorting logic so the view we're presenting the user seems consistent. (I _think_ the time we're currently sorting by is the _time the request was made_?)

**Comment by jonbell-lot23** [comment]
**Design status/point of view before I mark this as design-complete!**

<img width="646" alt="image" src="https://user-images.githubusercontent.com/9154902/168713450-402bd6a9-5e87-4b03-b45a-d641f3bbe2b8.png">

(this visual isn't pixel perfect yet, but it's a strong indication of where we're heading)

1. We're going to use a standard right-click menu that we use elsewhere
2. Options for "Set focus start" and "Set focus end"
3. Custom icons (not cut yet)
4. Unicorn palette

--

Some other details:

* We eventually want to have a "clear focus mode" option -- we can do that later
* I'd prefer to ship the focus mode stuff separately from the unicorn palette because there are bound to be UX considerations/nuances that come up. We can ship unicorn palette five minutes later, I just want to give each feature its own PR :)
* We should start with this in console and network monitor, then identify other places to address next

**Comment by bvaughn** [comment]
Since the design ticket has been closed, I guess the  focus mode portion of this is ready to be picked up for https://github.com/RecordReplay/devtools/issues/6735, so I'll do it today.
